### ISSUE TITLE
Undercurl not working in alacritty

### ISSUE THREAD

**Issue by asmodeus812** [neovim/neovim#21182]
### Neovim version (nvim -v)

0.8.1

### Vim (not Nvim) behaves the same?

NO

### Operating system/version

WSL2

### Terminal name/version

alacritty

### $TERM environment variable

alacritty

### Installation

github release

### How to reproduce the issue

**Having created the following file named `curl.vim` containing**
```vim
let &t_Cs = "\e[4:3m"
let &t_Ce = "\e[4:0m"

hi SpellBad     gui=undercurl guisp=red term=undercurl cterm=undercurl
```
**Start nvim with nvim --clean -u curl.vim (nvim version 0.8.1)** Scriptnames list below
![image](https://user-images.githubusercontent.com/15955811/203852349-24651b06-4d12-412e-b0fd-bae8ae8e7639.png)
![image](https://user-images.githubusercontent.com/15955811/203853115-6bb850c1-10f7-4b27-b3fb-6c158a9f65b3.png)

**Start vim with vim --clean -u curl.vim (vim version 8.1)**. Scriptnames list below
![image](https://user-images.githubusercontent.com/15955811/203852316-3e126521-cd2c-47e9-9545-3746edd1a08b.png)
![image](https://user-images.githubusercontent.com/15955811/203853059-7b804011-4dae-4fe1-8233-6386c0450d91.png)

**Alacritty behaviour with undercurls**
![image](https://user-images.githubusercontent.com/15955811/203852666-9561818a-c1ed-4fc9-a5c0-aa2674b8940e.png)

For some reason, the term is not being set on the nvim hi side, unlike vim which correctly sets that. Set termguicolors produces same behaviour - no undercurls for nvim

### Expected behavior

Underculrs are present in nvim

### Actual behavior

Undercurls are not visible in nvim, but are clearly visible in vim.

**Comment by zeertzjq** [comment]
Have you installed ncurses in WSL?

**Comment by asmodeus812** [comment]
I hadn't, i just tried doing `sudo apt-get install libncurses5-dev libncursesw5-dev` but no effect.

**Comment by zeertzjq** [comment]
Also try installing `ncurses-bin` and running `infocmp -x` in alacritty. What does that print?

**Comment by asmodeus812** [comment]
```bash
Reading package lists... Done
Building dependency tree
Reading state information... Done
ncurses-bin is already the newest version (6.2-0ubuntu2).
````

```
alacritty|alacritty terminal emulator,
	OTbs, am, bce, ccc, hs, mc5i, mir, msgr, npc, xenl, AX, XT,
	colors#0x100, cols#80, it#8, lines#24, pairs#0x10000,
	acsc=``aaffggiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
	bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
	clear=\E[H\E[2J, cnorm=\E[?12l\E[?25h, cr=\r,
	csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
	cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
	cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\E[A,
	cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
	dl=\E[%p1%dM, dl1=\E[M, dsl=\E]2;\007, ech=\E[%p1%dX,
	ed=\E[J, el=\E[K, el1=\E[1K, flash=\E[?5h$<100/>\E[?5l,
	fsl=^G, home=\E[H, hpa=\E[%i%p1%dG, ht=^I, hts=\EH,
	ich=\E[%p1%d@, il=\E[%p1%dL, il1=\E[L, ind=\n,
	indn=\E[%p1%dS,
	initc=\E]4;%p1%d;rgb\:%p2%{255}%*%{1000}%/%2.2X/%p3%{255}%*%{1000}%/%2.2X/%p4%{255}%*%{1000}%/%2.2X\E\\,
	invis=\E[8m, is2=\E[!p\E[?3;4l\E[4l\E>, kDC=\E[3;2~,
	kEND=\E[1;2F, kHOM=\E[1;2H, kIC=\E[2;2~, kLFT=\E[1;2D,
	kNXT=\E[6;2~, kPRV=\E[5;2~, kRIT=\E[1;2C, kb2=\EOE, kbs=^?,
	kcbt=\E[Z, kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC, kcuu1=\EOA,
	kdch1=\E[3~, kend=\EOF, kent=\EOM, kf1=\EOP, kf10=\E[21~,
	kf11=\E[23~, kf12=\E[24~, kf13=\E[1;2P, kf14=\E[1;2Q,
	kf15=\E[1;2R, kf16=\E[1;2S, kf17=\E[15;2~, kf18=\E[17;2~,
	kf19=\E[18;2~, kf2=\EOQ, kf20=\E[19;2~, kf21=\E[20;2~,
	kf22=\E[21;2~, kf23=\E[23;2~, kf24=\E[24;2~,
	kf25=\E[1;5P, kf26=\E[1;5Q, kf27=\E[1;5R, kf28=\E[1;5S,
	kf29=\E[15;5~, kf3=\EOR, kf30=\E[17;5~, kf31=\E[18;5~,
	kf32=\E[19;5~, kf33=\E[20;5~, kf34=\E[21;5~,
	kf35=\E[23;5~, kf36=\E[24;5~, kf37=\E[1;6P, kf38=\E[1;6Q,
	kf39=\E[1;6R, kf4=\EOS, kf40=\E[1;6S, kf41=\E[15;6~,
	kf42=\E[17;6~, kf43=\E[18;6~, kf44=\E[19;6~,
	kf45=\E[20;6~, kf46=\E[21;6~, kf47=\E[23;6~,
	kf48=\E[24;6~, kf49=\E[1;3P, kf5=\E[15~, kf50=\E[1;3Q,
	kf51=\E[1;3R, kf52=\E[1;3S, kf53=\E[15;3~, kf54=\E[17;3~,
	kf55=\E[18;3~, kf56=\E[19;3~, kf57=\E[20;3~,
	kf58=\E[21;3~, kf59=\E[23;3~, kf6=\E[17~, kf60=\E[24;3~,
	kf61=\E[1;4P, kf62=\E[1;4Q, kf63=\E[1;4R, kf7=\E[18~,
	kf8=\E[19~, kf9=\E[20~, khome=\EOH, kich1=\E[2~,
	kind=\E[1;2B, kmous=\E[<, knp=\E[6~, kpp=\E[5~,
	kri=\E[1;2A, mc0=\E[i, mc4=\E[4i, mc5=\E[5i, meml=\El,
	memu=\Em, oc=\E]104\007, op=\E[39;49m, rc=\E8,
	rep=%p1%c\E[%p2%{1}%-%db, rev=\E[7m, ri=\EM,
	rin=\E[%p1%dT, ritm=\E[23m, rmacs=\E(B, rmam=\E[?7l,
	rmcup=\E[?1049l\E[23;0;0t, rmir=\E[4l, rmkx=\E[?1l\E>,
	rmm=\E[?1034l, rmso=\E[27m, rmul=\E[24m,
	rs1=\Ec\E]104\007, rs2=\E[!p\E[?3;4l\E[4l\E>, sc=\E7,
	setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m,
	setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
	sgr=%?%p9%t\E(0%e\E(B%;\E[0%?%p6%t;1%;%?%p5%t;2%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m,
	sgr0=\E(B\E[m, sitm=\E[3m, smacs=\E(0, smam=\E[?7h,
	smcup=\E[?1049h\E[22;0;0t, smir=\E[4h, smkx=\E[?1h\E=,
	smm=\E[?1034h, smso=\E[7m, smul=\E[4m, tbc=\E[3g,
	tsl=\E]2;, u6=\E[%i%d;%dR, u7=\E[6n,
	u8=\E[?%[;0123456789]c, u9=\E[c, vpa=\E[%i%p1%dd,
	Cr=\E]112\007, Cs=\E]12;%p1%s\007, E3=\E[3J,
	Ms=\E]52;%p1%s;%p2%s\007, Se=\E[0 q, Ss=\E[%p1%d q,
	TS=\E]2;, XM=\E[?1006;1000%?%p1%{1}%=%th%el%;,
	kDC3=\E[3;3~, kDC4=\E[3;4~, kDC5=\E[3;5~, kDC6=\E[3;6~,
	kDC7=\E[3;7~, kDN=\E[1;2B, kDN3=\E[1;3B, kDN4=\E[1;4B,
	kDN5=\E[1;5B, kDN6=\E[1;6B, kDN7=\E[1;7B, kEND3=\E[1;3F,
	kEND4=\E[1;4F, kEND5=\E[1;5F, kEND6=\E[1;6F,
	kEND7=\E[1;7F, kHOM3=\E[1;3H, kHOM4=\E[1;4H,
	kHOM5=\E[1;5H, kHOM6=\E[1;6H, kHOM7=\E[1;7H,
	kIC3=\E[2;3~, kIC4=\E[2;4~, kIC5=\E[2;5~, kIC6=\E[2;6~,
	kIC7=\E[2;7~, kLFT3=\E[1;3D, kLFT4=\E[1;4D,
	kLFT5=\E[1;5D, kLFT6=\E[1;6D, kLFT7=\E[1;7D,
	kNXT3=\E[6;3~, kNXT4=\E[6;4~, kNXT5=\E[6;5~,
	kNXT6=\E[6;6~, kNXT7=\E[6;7~, kPRV3=\E[5;3~,
	kPRV4=\E[5;4~, kPRV5=\E[5;5~, kPRV6=\E[5;6~,
	kPRV7=\E[5;7~, kRIT3=\E[1;3C, kRIT4=\E[1;4C,
	kRIT5=\E[1;5C, kRIT6=\E[1;6C, kRIT7=\E[1;7C, kUP=\E[1;2A,
	kUP3=\E[1;3A, kUP4=\E[1;4A, kUP5=\E[1;5A, kUP6=\E[1;6A,
	kUP7=\E[1;7A, rmxx=\E[29m, smxx=\E[9m,
	xm=\E[<%i%p3%d;%p1%d;%p2%d;%?%p4%tM%em%;,
```

**Comment by xulongwu4** [comment]
I can reproduce this issue. I also noticed that, undercurl works in the alacritty terminal when running `echo -e "\e[4:3mTEST"` from the command line. However, if I enter neovim and then run `:term`, and then run `echo -e "\e[4:3mTEST"` in the opened terminal, undercurl does not work any more. This behavior does not happen when I tried kitty.

**Comment by xulongwu4** [comment]
I was able to fix the problem by installing the terminfo file provided in the alacritty release. It turns out that the terminfo file on my system was provided by an outdated ncurses (6.2) library and it needs to be updated.

**Comment by JaavierR** [comment]
@asmodeus812 I'm running alacritty from windows to connect to WSL2, but I can't get it to show the undercurl, I also specified the $TERM variable to be alacritty (after installing the alacritty terminfo file in WSL2) and by doing that I lose the colours of the folders or executables when doing `ls`, and the terminal doesn't even show an underline.

The problems are both in the terminal and in neovim, so I wanted to ask you also if the picture that shows the undercurl correctly is from alacritty in windows and if so, how did you configure it? because as I mentioned I have not been able to show the undercurl and I think I have the same configuration that you mention in the issue.

**Comment by asmodeus812** [comment]
@JaavierR if i remember correctly the image was from alacritty installed in wsl (ubuntu 22) not windows, the windows version of alacritty  does not support  under curls at all, i haven't had a chance to look back at this issue, but i remember that the wsl experience (it was somewhat sluggish) with alacritty was not the best, that is why i use the native .exe and connect to the wsl shell instead of starting alacritty from wsl.

**Comment by Syphdias** [comment]
I also have no undercurls in alacritty but not for the lack of support in alacritty as far as I can tell. I am just not sure if this is an alacritty issue, a neovim issue, or a catppuccin issue.

Short overview over my stack:
- (ArchLinux Kernel 6.6.8-arch1-1)
- (i3 4.23)
- alacritty 0.13.0 (78fa4d6f)  (undercurls do not work)
- kitty 0.31.0  (undercurls work)
- nvim v0.9.4
- LazyVim
- catppuccin

Alacritty sets `TERM=alacritty` while kitty sets `TERM=xterm-kitty`.
In Alacritty:
- If you set `TERM=alacritty` (default) the undercurls do not work.
- If you set `TERM=xterm-256color` the undercurls do not work.
- If you set `TERM=xterm-kitty` in alacritty the undercurls work.

I am not sure who handles that or what gets triggered or where it needs to be changed.
For now my workaround is to configure alacritty config to include:
```toml
[env]
TERM = "xterm-kitty"
```
Maybe this helps to find the core issue.

**Comment by zeertzjq** [comment]
Hmm, I tested the same Alacritty and Nvim versions on Arch Linux with `TERM=alacritty` and undercurls do work.

**Comment by Syphdias** [comment]
Hm, interesting. Maybe I am doing something wrong in my testing methodology?
```
alacritty --config-file /dev/null
echo $TERM #alacritty
nvim --clean
# type "asdf"
:se spell
:hi SpellBad cterm=undercurl
:q!
# this yields first red background for "asdf" then an additional red underline (not curled)

TERM=xterm-kitty nvim --clean
# type "asdf"
:se spell
:hi SpellBad cterm=undercurl
# this yields first red backgroudn for "asdf" then an additional red undercurl
```

**Comment by zeertzjq** [comment]
I tested the same steps. Undercurl works with `TERM=alacritty`.

**Comment by Syphdias** [comment]
Hm, I have no idea what could be differerent. I wrote `export` to two files to compare them and only show differences.
```
❯ comm -3 <(sort /tmp/kitty) <(sort /tmp/alacritty)
	ALACRITTY_LOG=/tmp/Alacritty-198367.log
	ALACRITTY_SOCKET=/run/user/1000/Alacritty-:0-198367.sock
	ALACRITTY_WINDOW_ID=44040197
KITTY_INSTALLATION_DIR=/usr/lib/kitty
KITTY_PID=197022
KITTY_PUBLIC_KEY='1:JcPdQQ=~<dGzA5p8EqnbJg~P3e~wT|0%P5+-LPsY'
KITTY_WINDOW_ID=1
	TERM=alacritty
TERM=xterm-kitty
	WINDOWID=44040197
WINDOWID=44040203
```
Any ideas where I could search?

**Comment by Syphdias** [comment]
I tried a few other terminal emulators:
- Tilix: `TERM=xterm-256color` undercurl works
- WezTerm: `TERM=xterm-256color` undercurl is a line
- Terminator:  `TERM=xterm-256color` undercurl works

**Comment by glepnir** [comment]
can't reproduce Undercurl works for me with alacritty.

**Comment by Syphdias** [comment]
Hi @zeertzjq and @glepnir, thank you for checking as well.
I think I solved it by understanding [this comment](https://github.com/neovim/neovim/issues/21182#issuecomment-1407429323).

I ran `infocmp -x` and noticed the first line:
```
❯ infocmp -x |head -2
#	Reconstructed via infocmp from file: /home/syphdias/.terminfo/a/alacritty
alacritty|alacritty terminal emulator,
```
I (re)moved the file but I did not "install"/copy the alacritty provided file — I don't think I need to.
```
❯ infocmp -x |head -2
#	Reconstructed via infocmp from file: /usr/share/terminfo/a/alacritty
alacritty|alacritty terminal emulator,
```
Starting a new alacritty instance did the trick and undercurl worked again for me.

I am thinking about removing the entire `~/.terminfo` directory to prevent this in the future. Not sure if it is needed if ncurses provides it in the `/usr/share/terminfo` already.

Sorry for the noise. I hope this will help someone who — like me — did not know about terminfo files :)

I also tried to fix it for WezTerm even though I do not use it currently.
(Re)moving `~/.terminfo/x/xterm-256color` did not fix the issue for WezTerm (eventhough `echo -e "\e[4:3mTEST"` does work outside of neovim). I found a comment from @wez about [the terminfo issue with nvim](https://github.com/wez/wezterm/issues/415#issuecomment-755849623) and used the command `curl https://raw.githubusercontent.com/wez/wezterm/master/termwiz/data/wezterm.terminfo | tic -x -` to install the wezterm terminfo.
I installed WezTerm through Flatpak which is the prefered way to install WezTerm, so I would argue it should be included in the packaging there. I think the [Arch Linux package includes the terminfo](https://gitlab.archlinux.org/archlinux/packaging/packages/wezterm/-/blob/07469716738429e1868bb382f04a8d0aa9410e2c/PKGBUILD#L92).

TL;DR: Fixed it by removing terminfo file from user directory: `mv ~/.terminfo/a/alacritty{,.bak}`

**Comment by gpanders** [comment]
>TL;DR: Fixed it by removing terminfo file from user directory: mv ~/.terminfo/a/alacritty{,.bak}

Just to clarify for future readers: the problem was not that the terminfo file was under `~/.terminfo`, but rather that in this case that particular terminfo file was old/out of date. Reinstalling the terminfo entry under `~/.terminfo` would have worked just as well (and may be necessary on systems where it's not possible to install terminfo entries into the system database).

**Comment by wez** [comment]
> I installed WezTerm through Flatpak which is the prefered way to install WezTerm

FWIW, It's not my preferred or recommended way to install wezterm. It's just convenient to try it for the first time.
Flatpak has a whole host of issues that stem from running in a sandbox; things like exporting terminfo from the container into the host are one of them. I'm not sure if that is even possible with flatpak.

I do recommend using native packages for your distro where possible.

**Comment by Syphdias** [comment]
Thanks for both clarifications!

Reference for installing terminfo file for alacritty: https://github.com/alacritty/alacritty/blob/master/INSTALL.md#terminfo

**Comment by ajoino** [comment]
I still have this issue, and none of the solutions in this thread seem to fit my situation:
 - Undercurls work in alacritty, but not in neovim.
 - My terminfo, using `infocmp -x` says alacritty and I don't have anything installed in `~/user/.terminfo`.
 - While `$TERM` is `alacritty` outside neovim, it changes to `xterm-256color` in neovim.
 - I've installed neovim from source.

What can I do to make undercurls work?

**Comment by gpanders** [comment]
@ajoino `$TERM` being `xterm-256color` in Neovim is fishy and is almost certainly the source of your problem. I'm not aware of anything in Neovim that would change your `$TERM` value, so it's likely something in your configuration/system (do you have a shell alias perhaps?)

**Comment by ajoino** [comment]
Thanks for the response @gpanders! I tested neovim in `gnome-terminal`+`bash` and it worked, but that's maybe because it uses the `xterm-256color` by default? My default is `alacritty`+`fish`.

I'll keep investigating tonight, but I'm getting the feeling my setup is a bit borked.

**Comment by ajoino** [comment]
So I read `h: terminfo`, followed the instructions to download the files and now undercurls work.
I still have some strangeness with `:!echo $TERM`=`alacritty` but in `:term` `echo $TERM` = `xterm-256color` but it looks right so I don't feel like investigating this further.

Turns out reading the helpful manual works!

**Comment by wookayin** [comment]
> `$TERM` being `xterm-256color` in Neovim is fishy and is almost certainly the source of your problem. I'm not aware of anything in Neovim that would change your `$TERM` value, so it's likely something in your configuration/system (do you have a shell alias perhaps?)

I can confirm that none of `~/.zshenv`, `~/.zshrc`, etc. sets `$TERM`, but `$TERM` is overwritten to `xterm-256color` even upon entering `~/.zshenv`. On neovim `vim.env.TERM` correctly prints `alacritty` or `wezterm`.

Indeed, this line sets `$TERM` upon `termopen`:

https://github.com/neovim/neovim/blob/v0.9.5/src/nvim/eval/funcs.c#L8821

Docs: `|termopen()|`

>		Terminal environment is initialized as in |jobstart-env|,
>		except $TERM is set to "xterm-256color". Full behavior is
>		described in |terminal|.

and it seems there's no way to prevent this being overwritten. Could we have this issue re-opened?

**Comment by clason** [comment]
I think it would be better to open a dedicated issue about this, arguing why this behavior is bad.

**Comment by clason** [comment]
(nobody wants to dig through dozens of tangential comments to get to the juicy and actionable pulp of the matter)

**Comment by gpanders** [comment]
> Indeed, this line sets `$TERM` upon `termopen` or `:term`:
> 
> https://github.com/neovim/neovim/blob/v0.9.5/src/nvim/eval/funcs.c#L8821
> 
> Docs: `|termopen()|`
> 
> > ```
> >   Terminal environment is initialized as in |jobstart-env|,
> >   except $TERM is set to "xterm-256color". Full behavior is
> >   described in |terminal|.
> > ```
> 
> and it seems there's no way to prevent this from being overwritten. Could we have this issue re-opened?

That describes the embedded terminal emulator only. It sets `$TERM` *inside* the embedded terminal, it does not modify `$TERM` for the parent Nvim process. 

You can set `TERM` to whatever you like using the `env` key to `termopen()`, but making it to default to the same value as the host terminal is not a good idea: the embedded terminal is libvterm, *not* Alacritty (or whatever).
